Level based outer approximation

ZedongPeng opened the pull request:
## Fixes # .

## Summary/Motivation:


## Changes proposed in this PR:
-
-

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

emma58 committed:
Adding integer arithmetic option and some tests of the dual for infinity norm

michaelbynum committed:
Pynumero: user should not have to call broadcast_block_sizes

michaelbynum committed:
Merge pull request #1663 from michaelbynum/derivs

updating methods for differentiating pyomo expressions

michaelbynum committed:
Pynumero: user should not have to call broadcast_block_sizes

michaelbynum committed:
Pynumero: user should not have to call broadcast_block_sizes

emma58 committed:
Adding handling for useless bigM variables: If a variable is stale after the first iteration solve, we will not include it in the separation problem. Also cleaning up imports in cuttingplane.py since I wasn't using half of them

emma58 committed:
Fixing imports in test_cuttingplane... It was going to have conflicts with Miranda's PR anyway. :(

emma58 committed:
NFCs--removing a lot of inadvertant changes and some unused additions

michaelbynum committed:
Pynumero: user should not have to call broadcast_block_sizes

michaelbynum committed:
removing old print statement

michaelbynum committed:
missing check for block structure

jsiirola committed:
Disable codecov S3 uploads on Jenkins

michaelbynum committed:
PyNumero: size should be an int

jsiirola committed:
Merge pull request #1670 from jsiirola/codecov-no-s3

Updating Jenkins driver to resolve codecov report upload

michaelbynum committed:
Merge branch 'master' into no_broadcast

blnicho committed:
Merge pull request #1661 from Moisan/encodestring_deprecation

Deprecate the use of base64.decodestring with python3

jsiirola committed:
PEP8 fix

jsiirola committed:
Restoring removed code as documentation comment

jsiirola committed:
Fixing bad merge with chull.py that accidentally reverted a recent commit

michaelbynum committed:
Merge pull request #1667 from michaelbynum/no_broadcast

PyNumero Updates

jsiirola committed:
Updating __init__.py so that imported symbols are unchanged

mrmundt committed:
Fixing merge conflicts

mrmundt committed:
Incorporating jsiirola's changes

jsiirola committed:
Add hook to Jenkins driver for triggering cythonization

jsiirola committed:
Add hook for cythonization; cythonized python 3.9 build

jsiirola committed:
Setting up a slim GHA build

jsiirola committed:
Fixing typo

jsiirola committed:
Strip newlines from env variable

jsiirola committed:
Replace newlines in ENV variable

jsiirola committed:
OSX compatibility

mrmundt committed:
Adding a bug message to test_odbc_ini

jsiirola committed:
Merge pull request #1674 from jsiirola/add-cython-hooks

Update GHA and Jenkins drivers to support slim and cythonized builds

mrmundt committed:
Fixing bug in test_odbc_init

jsiirola committed:
Make Constraint.Feasible be an alias for Constraint.Skip

jsiirola committed:
Fixing typo in base class name

mrmundt committed:
Applying bug fix to mindtpy, as per @bernalde

jsiirola committed:
Defer import of profiling packages

mrmundt committed:
Fixing whitespaces and newlines

mrmundt committed:
Applying bug fix for units_container TypeError

mrmundt committed:
Merge branch 'master' into import_fixes

mrmundt committed:
One more typo - and then done.

jsiirola committed:
Bundle the GHA coverage reports into fewer uploads

jsiirola committed:
Fixing YML syntax error

jsiirola committed:
Set codecov script as an environment var

jsiirola committed:
Adding debugging

jsiirola committed:
Upload coverage even if a build fails

jsiirola committed:
Setup python, install coverage for the codecov upload job

jsiirola committed:
Synchronizing GHA workflows

jsiirola committed:
Updating expected number of coverage reports

jsiirola committed:
Add a fallback to upload coverage 4.x results

jsiirola committed:
Fix bash syntax error

jsiirola committed:
Adjusting expected number of build reports

ZedongPeng committed:
add skip_stale_vars option for store_to function

jsiirola committed:
Additional build log inforamtion

jsiirola committed:
Move coverage artifacts to the root dir before combining

jsiirola committed:
Fix typo, add debugging

jsiirola committed:
Fix coverage combination; add additional debugging

jsiirola committed:
Add missing source checkout

jsiirola committed:
Create parse_table_datacmds, general cleanup

blnicho committed:
Merge pull request #1676 from jsiirola/constraint-feasible-fix

Make Constraint.Feasible be an alias for Constraint.Skip

mrmundt committed:
Merge remote-tracking branch 'origin' into import_fixes

jsiirola committed:
Merge pull request #1677 from jsiirola/profile-imports

Defer import of profiling packages

mrmundt committed:
Merge branch 'master' into import_fixes

jsiirola committed:
install pyutilib

jsiirola committed:
Remove unnecessary mkdir

jsiirola committed:
Merge pull request #1638 from Robbybp/slice-bug

Allow comparison between slices with one-dimensional slices in __getitem__ call

jsiirola committed:
split coverage processing into os-specific jobs

jsiirola committed:
Correct pattern for target coverage files

jsiirola committed:
synchronizing GHA workflows

jsiirola committed:
Adding tags to the GHA codecov uploads

jsiirola committed:
Switch format for codecov arguments

jsiirola committed:
Merge pull request #1680 from jsiirola/bundle-gha-cov-reports

Bundle GHA coverage report uploads

jsiirola committed:
Update pynumero's mumps interface to use attempt_import

jsiirola committed:
Add a Pyomo-specific unittest.TestCase wrapper

This wrapper implements a new assertRelativeEqual() for comparing
structured data (scalars, sequences, and mappings) with a relative
tolerance

jsiirola committed:
Remove 'Time' when comparing solver results

jsiirola committed:
Remove debugging

jsiirola committed:
Merge branch 'master' into fme-var-bounds

jsiirola committed:
Rename pyomo.common.testing so nose doesn't think it is a test

jsiirola committed:
Merge pull request #1664 from emma58/fme-var-bounds

Treat bounds of of variables being projected as constraints in FME transformation

jsiirola committed:
Merge branch 'master' into improve_store_to

jsiirola committed:
Update attempt_import to support deferred import of submodules

jsiirola committed:
Add additional *_available flags to the RTD configuration

jsiirola committed:
Update parmest.graphics to use deferred module/submodule imports

ZedongPeng committed:
Merge branch 'mindtpy-master' into level-based-outer-approximation

ZedongPeng committed:
Merge branch 'mindtpy-master' into level-based-outer-approximation

ZedongPeng committed:
create setup_master function

ZedongPeng committed:
halfway LOA

emma58 committed:
Adding option to specify the name of the projected constraints in the FME transformation

emma58 committed:
Addressing all of John's comments except for the exact equality issue. Also adding option to name the cuts so that they are accessible outside of the transformation block.

ZedongPeng committed:
fix objective activate and deactivate issue

ZedongPeng committed:
first working version of LOA

ZedongPeng committed:
update logger

ZedongPeng committed:
move del_component('loa_proj_mip_obj')

ZedongPeng committed:
fix L1, L2, L inf objective variable scope

ZedongPeng committed:
Revert "fix L1, L2, L inf objective variable scope"

This reverts commit 1d0ee1d4b37f2161f923ae842402ec3e33063b89.

ZedongPeng committed:
fix L1,L2,L-inf objective variable scope

emma58 committed:
Including exactly satisfied equalities in FME, and adding a test of this. Also, precomputing the rHull constraints that involve disaggregated variables as those are the only ones that could be useful.

emma58 committed:
Resolving conflicts in FME tests

mrmundt committed:
Adding @blnicho's suggestions

mrmundt committed:
Reverting opt/plugins/sol.py change

jetuk committed:
fix-ph-convert-string-to-number

The comment `ph.py` suggests that an attempt is made to convert string arguments to numbers. It says that integer is tried first, and then float. However, the implementation is opposite and consequently all integer arguments are converted to floats. This has caused a problem in my CBC model when I try to pass an integer value.

The changes here switch the order of the conversion attempts to reflect the written comment and allow integer arguments to be passed correctly.

blnicho committed:
Merge pull request #1607 from mrmundt/import_fixes

Import Statement Removal and Replacements

blnicho committed:
Merge pull request #1684 from jetuk/patch-1

Fix integer conversion of string arguments in ProgressiveHedging

michaelbynum committed:
Merge pull request #1681 from jsiirola/pynumero-mumps-import

Update pynumero's mumps interface to use attempt_import

emma58 committed:
Fixing merge conflicts with import statements

ZedongPeng committed:
fix bug in LOA

jsiirola committed:
Merge pull request #1683 from jsiirola/dependencies-submodules

Add support for deferred import of submodules

ZedongPeng commented:
I create a new PR from `level-based-outer-approximation` branch to `mindtpy-master`, instead of `master`. So only the updates about LOA method will be shown here. @bernalde

mrmundt committed:
Moving pyutilib.misc.config into pyomo.common.config

mrmundt committed:
Moving pyutilib.misc.timing into pyomo.common.timing

ZedongPeng committed:
set skip_stale_vars=False by default

ZedongPeng committed:
add test of store_to() with skip_stale_vars=True

bernalde commented on pyomo/contrib/mindtpy/MindtPy.py:
```diff
@@ -109,10 +109,12 @@ def solve(self, model, **kwds):
             config.init_strategy = "feas_pump"
             config.iteration_limit = 0
             config.add_no_good_cuts = True
+        elif config.strategy == "LOA":
```
This is the first and main part where we should discuss it. Should LOA be a strategy different from OA or ECP, or rather a different option that, when True, makes you solve the extra MIP subproblems? I think it makes more sense in the second case.

bernalde commented on pyomo/contrib/mindtpy/config_options.py:
```diff
@@ -147,6 +148,7 @@ def _get_MindtPy_config():
         domain=bool
     ))
     CONFIG.declare("use_dual", ConfigValue(
+        # TODO: should we set the default value to True?
```
I think that we rename this option into `equality_relaxation` option that controls adding the duals. Both this option and `use_slacks` can be controlled by a larger option called `heuristic_nonconvex`

bernalde commented on pyomo/contrib/mindtpy/config_options.py:
```diff
@@ -330,3 +333,11 @@ def _add_feas_pump_configs(CONFIG):
+    CONFIG.declare("loa_coef", ConfigValue(
+        default=0.5,
+        domain=PositiveFloat,
+        description="loa_coef"
```
I think we need a better description here.

bernalde commented on pyomo/contrib/mindtpy/mip_solve.py:
```diff
@@ -388,3 +339,89 @@ def handle_master_unbounded(master_mip, solve_data, config):
+                solve_data.working_model,
+                discrete_only=config.fp_discrete_only)
+    elif loa_projection:
+        MindtPy.loa_proj_mip_obj = generate_norm2sq_objective_function(solve_data.mip,
```
This is the point where our linear approximation may play a role, right?

ZedongPeng replied:
Yes

bernalde commented on pyomo/contrib/mindtpy/nlp_solve.py:
```diff
@@ -247,7 +247,7 @@ def handle_subproblem_infeasible(fixed_nlp, solve_data, config):
                              solve_data.mip.MindtPy_utils.variable_list,
                              config)
-        if config.strategy == "OA":
+        if config.strategy in {"OA", "LOA"}:
```
We should use either " " or ' ' uniformly in the code.

ZedongPeng replied:
I'll handle this after we merge all the PRs. Otherwise, it'll bring us a lot of trouble when merging different branches.

bernalde commented on pyomo/contrib/mindtpy/tests/test_mindtpy_LOA.py:
```diff
@@ -0,0 +1,311 @@
+
+    #         self.assertIs(results.solver.termination_condition,
+    #                       TerminationCondition.optimal)
+    #         self.assertAlmostEqual(value(model.cost.expr), 3.5, places=2)
```
Many tests are commented on here. I do not know if this is useful. Either uncomment them all or remove them.

bernalde commented on pyomo/contrib/mindtpy/util.py:
```diff
@@ -197,8 +197,10 @@ def generate_norm2sq_objective_function(model, setpoint_model, discrete_only=Fal
     discrete_only: Bool
         only optimize on distance between the discrete variables
     """
+    # TODO: do we need to skip config.add_slack variable here?
```
As per our last meeting, yes.

bernalde replied:
Replace the TODO with a comment and our answer (yes)

bernalde reviewed (changes requested):
Hi Zedong, great work again. The main change that I would like to discuss is to add 2 extra options:
1. `add_regularization `with different cases. `level_linear`, `level_squared` (this replaces your current `LOA `implementation), `grad_lag`, and `hess_lag `(the names for the last two we can still discuss). This option should enable the solution of an extra MI(Q)P after each MILP once a feasible solution has been found, and should be independent of the `strategy` chosen.
2. `heur_nonconvex `that activates `compute_duals `and `add_slacks`
We can talk this through with more detail but I think that this would allow for some pretty new combinations of algorithms. Think using feaspump to find a feasible solution and then LOA but with ECP or OA for example.

mrmundt committed:
First attempt at moving tests

mrmundt committed:
os.environ call correction

mrmundt committed:
Moving pyutilib.common exceptions to pyomo.common.errors

mrmundt committed:
Incorporating first round of suggestions

ZedongPeng commented:
> Hi Zedong, great work again. The main change that I would like to discuss is to add 2 extra options:
> 
> 1. `add_regularization `with different cases. `level_linear`, `level_squared` (this replaces your current `LOA `implementation), `grad_lag`, and `hess_lag `(the names for the last two we can still discuss). This option should enable the solution of an extra MI(Q)P after each MILP once a feasible solution has been found, and should be independent of the `strategy` chosen.
> 2. `heur_nonconvex `that activates `compute_duals `and `add_slacks`
>    We can talk this through with more detail but I think that this would allow for some pretty new combinations of algorithms. Think using feaspump to find a feasible solution and then LOA but with ECP or OA for example.

I agree that using `add_regularization` is more reasonable. What does `level_linear` mean here?

bernalde commented:
LOA is level based outer approximation according to our paper with Jan. The linear level OA is what you and I are working on right now! (Norm 1 or infinity in the auxiliary MIP)

ZedongPeng committed:
verified LOA

ZedongPeng committed:
repalce LOA with add_regularization

ZedongPeng committed:
rename use_dual to equality_relaxation

ZedongPeng committed:
add heuristic_nonconvex

ZedongPeng committed:
fix cycling check bug in LOA

ZedongPeng committed:
modify LOA test

ZedongPeng committed:
add condition for calculating tmp_dual, remove redundant code

ZedongPeng committed:
update description of loa_coef and rename it to level_coef

mrmundt committed:
Removing unnecessary import statement

jsiirola committed:
Fix relative comparisons; distinguish between string and sequence

jsiirola committed:
Add tests for unittest module

jsiirola committed:
defining pyomo.common.unittest.__all__

jsiirola committed:
Correcting typos in comments

jsiirola committed:
Merge pull request #1682 from jsiirola/testing-assertrelativeequal

Create Pyomo-specific unittest.TestCase and update fragile test

ZedongPeng commented:
Hey @bernalde . There are also some updates in this PR. Give it a look if you have free time.

ZedongPeng committed:
add test_store_to_with_skip_stale_vars

ZedongPeng committed:
modify test function

emma58 committed:
John's suggestion and somehow the FME tests didn't register the hull transformation

emma58 committed:
Merge branch 'master' into cuttingplanes-rewrite

emma58 committed:
Accomodating manually named constraints in the FME post-processing function

emma58 committed:
Switching to ConfigBlock for solver options

mrmundt committed:
Creating pyomo.common.factory

mrmundt committed:
Update factory.py

mrmundt committed:
Moving pyutilib.factory (Factory) to pyomo.common.factory

mrmundt committed:
Can't import something that doesn't exist

bernalde commented on pyomo/contrib/mindtpy/config_options.py:
```diff
@@ -36,6 +36,13 @@ def _get_MindtPy_config():
+        default=None,
+        domain=In(["level_linear", "level_squared", "grad_lag", "hess_lag"]),
+        description="add regularization",
+        doc="solving a projection problem before solve the fixed subpoblem"
```
subproblem typo

bernalde commented on pyomo/contrib/mindtpy/initialization.py:
```diff
@@ -134,17 +139,18 @@ def init_rNLP(solve_data, config):
+            m.dual[c] for c in MindtPy.constraint_list) if config.equality_relaxation else None
         # Add OA cut
         # This covers the case when the Lower bound does not exist.
         # TODO: should we use the bound of the rNLP here?
```
Only if available.

bernalde commented on pyomo/contrib/mindtpy/mip_solve.py:
```diff
@@ -43,76 +43,24 @@ def solve_master(solve_data, config, feas_pump=False):
         result from solving the master MIP
+    feas_pump: Bool
+        generate the feasibility pump projection master problem
+    loa_projection: Bool
```
Maybe change this argument name to `regularization_problem` instead

bernalde commented on pyomo/contrib/mindtpy/tests/MINLP4_simple.py:
```diff
@@ -0,0 +1,37 @@
+        m.c2 = Constraint(expr=1/m.x + 1/m.y - sqrt(m.x) * sqrt(m.y) <= -1)
+        m.c3 = Constraint(expr=2*m.x - 5*m.y <= -1)
+
+        m.obj = Objective(expr=-6 * m.x - m.y, sense=minimize)
```
Add new line at end of file

bernalde commented on pyomo/contrib/mindtpy/tests/test_mindtpy.py:
```diff
@@ -33,18 +33,34 @@
         """Test the outer approximation decomposition algorithm."""
         with SolverFactory('mindtpy') as opt:
-            model = EightProcessFlowsheet(convex=False)
+            model = EightProcessFlowsheet()
```
Let's be explicit here and add argument `convex=True` even if it is already on by default.

bernalde commented on pyomo/contrib/mindtpy/tests/test_mindtpy_LOA.py:
```diff
@@ -0,0 +1,223 @@
+# -*- coding: utf-8 -*-
```
Change the name to test regularization

bernalde reviewed (changes requested):
Few changes requested, this is in really good shape. Let's start working on the Hessian of the Lagrangean

jsiirola committed:
Merge pull request #1668 from emma58/cuttingplanes-rewrite

Cuttingplanes rewrite!

ZedongPeng committed:
fix typo

ZedongPeng committed:
rename loa_projection to regularization_problem

ZedongPeng committed:
add new line

ZedongPeng committed:
add convex=True for EightProcessFlowsheet

ZedongPeng committed:
rename test_mindtpy_LOA to test_mindtpy_regularization

ZedongPeng committed:
resolve TODO

mrmundt committed:
Bumping miniconda to v2 to apply get-path fix

mrmundt committed:
Removing deprecated portion

jsiirola committed:
Merge pull request #1696 from mrmundt/conda-fix

Changing setup-miniconda action to version 2

jsiirola committed:
Defer import of guppy module within PySP

jsiirola committed:
Merge pull request #1697 from jsiirola/pysp-guppy-import

Defer import of guppy module within PySP

ZedongPeng committed:
remove norm block after feasibility pump

ZedongPeng committed:
fix file name

ZedongPeng committed:
add level_L1 and level_L_infinity

blnicho committed:
Removing log file that was committed by mistake

mrmundt committed:
Merge branch 'master' of https://github.com/Pyomo/pyomo into pyutilib-factory

mrmundt committed:
Merge remote-tracking branch 'origin' into pyutilib-common

mrmundt committed:
Changing import statement to not truly divorce pyutilib for now

blnicho committed:
Merge pull request #1679 from ZedongPeng/improve_store_to

Add skip_stale_vars option for store_to function

jsiirola committed:
GHA: always pull down the most recent build of IDAES-ext

jsiirola committed:
Bumping GHA cache version

jsiirola committed:
Bumping GHA cache action version

jsiirola committed:
Fix Python 2.7, 3.4 compatibility

jsiirola committed:
PEP8 fixes

jsiirola committed:
Merge pull request #1698 from jsiirola/gha-update-idaes-download

GHA: update idaes-ext download

mrmundt committed:
Incorporating changes - but they might be unnecessary since we might remove this file

ZedongPeng committed:
fix bug

blnicho committed:
Merge pull request #1699 from jsiirola/parmest-compatibility

Resolve Parmest example incompatibility with Python<3.5

mrmundt committed:
Adding copyright

jsiirola committed:
Merge pull request #1695 from mrmundt/pyutilib-factory

Move pyutilib.factory to pyomo.common.factory

mrmundt committed:
Adding tests for timing

mrmundt committed:
Reverting phrasing changes

jsiirola committed:
Improve robustness of PseudoMap _sorted restoration

jsiirola committed:
Improve Block initialization

- switch Block to leverage Initializer()
- standardize initialization in _getitem_when_not_present
- resolve edge cases for constructing nested abstract blocks
- improve testing
- deprecate the Block(options=...) argument

jsiirola committed:
Switch block.py to use the deprecated decorator

jsiirola committed:
import Mapping from pyomo.common.collections

jsiirola committed:
Simplify the definition of SimpleBlock.display

jsiirola committed:
Add a 'dense' argument to Block()

michaelbynum committed:
assert correct owners should not be required

michaelbynum committed:
Merge branch 'master' into pynumero

carldlaird committed:
Update mpi_block_matrix.py

Adding documentation for assert_correct_owners

jsiirola committed:
Merge pull request #1694 from mrmundt/pyutilib-common

Move pyutilib.common exception references to pyomo.common.errors

jsiirola committed:
Relaxing some test tolerances to reflect changes in xpress 8.11

michaelbynum committed:
Merge pull request #1706 from jsiirola/xpress-test-updates

Resolve test xpress interface test failures with xpress 8.11

michaelbynum committed:
Merge branch 'master' into pynumero

michaelbynum committed:
Merge remote-tracking branch 'michaelbynum/pynumero' into pynumero

blnicho committed:
Fixing typo in a comment

michaelbynum committed:
Merge pull request #1705 from michaelbynum/pynumero

Pynumero: MPIBlockMatrix: assert correct owners should not be required

jsiirola committed:
Remove '__all__'

jsiirola committed:
Update tic/toc so that msg=None suppresses stream output

jsiirola committed:
Merge branch 'master' into pyutilib-timing

jsiirola committed:
Improve TicTicTimer handling of 'not specified' osstream and logger

jsiirola committed:
Improve timing tests coverage and robustness

jsiirola committed:
Rearranging timing test to help resolve failure on pypy

jsiirola committed:
Resolve nondeterminism in str(HierarchicalTimer)

jsiirola committed:
Relax timing tolerance to make pypy tests more robust

jsiirola committed:
Another attempt to resolve pypy2 timing test failures

jsiirola committed:
Remove trailing whitespace

jsiirola committed:
Cleanup tests: whitespace, imports

jsiirola committed:
Config: cleanup imports, whitespace; standardize on ConfigDict name

jsiirola committed:
config: cleanup/update tests (whitespace, imports, and exceptions)

jsiirola committed:
Merge branch 'master' into pyutilib-config

jsiirola committed:
Update exception string regex for recent Python builds

jsiirola committed:
Tracking change in exception string

ZedongPeng committed:
first working version of grad_lag and  hess_lag

Moisan committed:
Add docstring to some classes in pyomo/common/config.py

michaelbynum committed:
Merge pull request #1703 from jsiirola/block-initialization

Block initialization improvement

ZedongPeng committed:
add more tests (regularization)

jsiirola committed:
Merge pull request #1688 from mrmundt/pyutilib-timing

Move pyutilib.misc.timing into pyomo.common.timing

mrmundt committed:
Merge branch 'master' of https://github.com/Pyomo/pyomo into pyutilib-config

mrmundt committed:
Removing argparse availability check

jsiirola committed:
Merge pull request #1687 from mrmundt/pyutilib-config

Move pyutilib.misc.config into pyomo.common.config

mrmundt committed:
Move pyomo.misc Bunch/Container/Options to pyomo.common.collections

mrmundt committed:
Safety-net: Changing to ubuntu-18.04 to protect suite

mrmundt committed:
Adding breadcrumb trail

mrmundt committed:
Fixing plural of index

mrmundt committed:
First round - does not work

mrmundt committed:
Second round - still doesn't work

mrmundt committed:
Third round

bernalde commented on pyomo/contrib/mindtpy/tests/test_mindtpy_regularization.py:
```diff
@@ -156,6 +156,8 @@ def test_OA_ConstraintQualificationExample_level_L1(self):
                           TerminationCondition.optimal)
             self.assertAlmostEqual(value(model.objective.expr), 3, places=2)
+    '''
+    # bug
```
What is this bug about?

ZedongPeng replied:
Bug fixed.

mrmundt committed:
Fourth round

mrmundt committed:
Fifth round

mrmundt committed:
Changing appropriate assertions to regex

mrmundt committed:
Changing limit

bernalde commented on pyomo/contrib/mindtpy/util.py:
```diff
@@ -288,6 +294,66 @@ def generate_norm_inf_objective_function(model, setpoint_model, discrete_only=Fa
+    #                        for i, var in enumerate(model.MindtPy_utils.variable_list[:-1]))
+    # second_order_term = 0.5 * sum((var_i - var_i.value) * float(hess_lag[i][j]) * (var_j - var_j.value)
+    #                               for i, var_i in enumerate(model.MindtPy_utils.variable_list[:-1])
+    #                               for j, var_j in enumerate(model.MindtPy_utils.variable_list[:-1]))
```
Can you maybe explain why is this second approach not used? How did it fail?

ZedongPeng replied:
It didn't fail. Both implementations work well. We just need to choose one of the two.

bernalde commented on pyomo/contrib/mindtpy/util.py:
```diff
@@ -288,6 +294,66 @@ def generate_norm_inf_objective_function(model, setpoint_model, discrete_only=Fa
+    if config.add_regularization == "grad_lag":
+        return Objective(expr=first_order_term, sense=minimize)
+    elif config.add_regularization == "hess_lag":
+        return Objective(expr=first_order_term + second_order_term, sense=minimize)
```
I would only construct the second-order term (calling the hessian of the lagrangian and writing the objective in memory) if this condition is satisfied. In the linear case, we do not want to go through all this trouble.

ZedongPeng replied:
Do you mean we only use Pynumero to calculate the second term? For the first term, we still use the jacobian in MindtPy?

bernalde replied:
No, I meant exactly what you did in the code.
`elif config.add_regularization == "hess_lag":`  Calculate the second-order things

bernalde commented on pyomo/contrib/mindtpy/util.py:
```diff
@@ -288,6 +294,66 @@ def generate_norm_inf_objective_function(model, setpoint_model, discrete_only=Fa
+    # TODO: PyNumero does not support discrete variables
+    # So PyomoNLP should operate on setpoint_model
+
+    # Implementation 1
```
This is very clean! Great job and amazing that you could use PyNumero in such a way. The issue is that we might be repeating some work here that we can save, for example when you call PyomoNLP you are generating the .nl file, which (in theory) should already exist from the solution of the NLP. Is this part "expensive" and should we try to spend some time recycling the .nl file that already exists or not?

ZedongPeng replied:
I find an interesting about the .nl file in Pyomo. If I call baron, the file generated is .bar. If I call gams, the file generated is .gms. If I call ipopt, the file generated is .nl.
How can we unify this with different solvers?

bernalde commented on pyomo/contrib/mindtpy/util.py:
```diff
@@ -288,6 +294,66 @@ def generate_norm_inf_objective_function(model, setpoint_model, discrete_only=Fa
+        return Objective(expr=first_order_term + second_order_term, sense=minimize)
+
+
 def generate_norm1_norm_constraint(model, setpoint_model, config, discrete_only=True):
```
The name of this function is misleading. We are generating the constraints to represent the Norm 1 objective. Try editing it as `generate_norm1_objective_constraint`

ZedongPeng replied:
The name of the constraint is `norm_constraint`, should we also change the name of the constraint?

bernalde commented on pyomo/contrib/mindtpy/iterate.py:
```diff
@@ -68,6 +68,15 @@ def MindtPy_iteration_loop(solve_data, config):
         else:
             raise NotImplementedError()
 
+        if config.add_regularization is not None:
```
Regularization should only be considered if there is an existing incumbent (upper bound) solution. This would remove a part of the conditional in the next line.

bernalde commented on pyomo/contrib/mindtpy/config_options.py:
```diff
@@ -161,11 +168,17 @@ def _get_MindtPy_config():
     ))
+    CONFIG.declare("heuristic_nonconvex", ConfigValue(
+        default=False,
+        description="use dual solution from the nlp solver and slack variables to add OA cuts for equality constraints.",
```
Replace with "use dual solution from the NLP solver and slack variables to add OA cuts for equality constraints (Equality relaxation) and minimize the sum of the slack variables (Augmented Penalty)"

bernalde reviewed (commented):
Really impressive that this is up and working. Besides some cosmetic changes and some minor performance improvements, this is only missing the computational results.

ZedongPeng committed:
remove bug

ZedongPeng committed:
update documentation

ZedongPeng committed:
config.add_regularization condition improvement

blnicho committed:
Merge pull request #1707 from Moisan/common_config_docstring

Add docstring to some classes in pyomo/common/config.py

mrmundt committed:
Adding more flexibility

mrmundt committed:
Removing front spaces

mrmundt committed:
Line by line instead of as a chunk

jsiirola committed:
Merge pull request #1709 from mrmundt/ubuntu-18

Replace ubuntu-latest with ubuntu-18.04 in GitHub Action Workflows

mrmundt committed:
Merge pull request #1708 from mrmundt/pyutilib-bco

Move pyomo.misc Bunch/Container/Options import to pyomo.common.collections

mrmundt committed:
Merge pull request #1711 from mrmundt/timing_test_fix

Make pyomo.common.timing tests more flexible

ZedongPeng committed:
used dual if grad_lag or hess_lag

ZedongPeng committed:
add objective_list

ZedongPeng committed:
improve MindtPy with objective_list and objective_sense

ZedongPeng committed:
fix bug

ZedongPeng committed:
remove redundant find_component

ZedongPeng committed:
only calculate second_order_term if hess_lag

ZedongPeng committed:
remove redundant import

bernalde commented on pyomo/contrib/mindtpy/MindtPy.py:
```diff
@@ -121,6 +115,8 @@ def solve(self, model, **kwds):
             config.mip_solver_tee = True
             config.nlp_solver_tee = True
+        if config.add_regularization in {'grad_lag', 'hess_lag'}:
+            config.equality_relaxation = True
```
Why do you need to do this? I don't think it is necessary since it should work without the equality relaxation.

ZedongPeng replied:
This is because the duals are only calculated when `equality_relaxation = True`. The old name of `equality_relaxation` is `use_dual`. Do you think we need a new option to obtain the duals of the NLPs?

bernalde replied:
What do you need the duals for? To construct the Lagrangian? I think that that should come from the PyomoNLP model instead of the original NLP model.

ZedongPeng replied:
I just tested it. If we don't define the dual suffix in the original NLP model, the generated PyomoNLP model doesn't have dual values (all zero, using `get_duals()` function). So I think we still need to get the duals from the Pyomo original model.

bernalde replied:
Perfectly reasonable. Do you think we should have separate options to calculate duals and for the equality relaxation? In a sense, the equality relaxation does much more than only compute the duals.

bernalde commented on pyomo/contrib/gdpopt/util.py:
```diff
@@ -294,6 +295,11 @@ def build_ordered_component_lists(model, solve_data):
+        util_blk, 'objective_list', list(
+            model.component_data_objects(
+                ctype=Objective, active=True,
+                descend_into=(Block))))
```
Can you add a comment for this? It appears as a component list for all the objectives that you are using in your problem.

ZedongPeng replied:
Yep. This a component list for all the objectives in the problem. Before solving the problem, we also check the number of objectives in the problem and we don't support the multi-objective problems.

https://github.com/Pyomo/pyomo/blob/08f4188bf5553a5aaf9856f9d5aa45ff76b4f675/pyomo/contrib/gdpopt/util.py#L134

So generally, the objective_list has only one component, except we add a new variable and move the objective to constraints. 
The reason behind adding the objective_list is that in our previous code is that there are so many places where we search for the active objective, like here 
https://github.com/ZedongPeng/pyomo/blob/45a1b1072cf3525db4adf807dfad5ae18cf45ac2/pyomo/contrib/mindtpy/mip_solve.py#L56

I want to make our code more efficient and more deterministic, so I add the objective_list and replace `main_objective=next(...)` with it.

bernalde replied:
Great improvement. I like it very much

bernalde commented on pyomo/contrib/mindtpy/feasibility_pump.py:
```diff
@@ -55,10 +49,8 @@ def solve_feas_pump_subproblem(solve_data, config):
-        fp_nlp.component_data_objects(Objective, active=True))
-    main_objective.deactivate()
-    if main_objective.sense == minimize:
+    fp_nlp.MindtPy_utils.objective_list[-1].deactivate()
```
`MindtPy_utils.objective_list[-1]` is the linear objective? If so why does it appear in `fp_nlp`?

ZedongPeng replied:
`fp_nlp` is first defined as `solve_data.working_model.clone()`. So it has everything as the `working_model`, including the objective function. I first deactivate the old objective function and then generate the new objective function.

bernalde commented on pyomo/contrib/mindtpy/feasibility_pump.py:
```diff
@@ -213,20 +200,21 @@ def feas_pump_loop(solve_data, config):
-    if solve_data.mip.MindtPy_utils.MindtPy_linear_cuts.find_component('improving_objective_cut') is not None:
-        solve_data.mip.MindtPy_utils.MindtPy_linear_cuts.improving_objective_cut.deactivate()
+    solve_data.mip.MindtPy_utils.MindtPy_linear_cuts.del_component(
+        'improving_objective_cut')
```
It seems as if you are starting to delete objectives instead of deactivating them. Does this mean that you need to create them again at each iteration?

ZedongPeng replied:
Here improving_objective_cut is a constraint. In Pyomo, if a constraint is defined, it's hard for us to change it. (I didn't find any good example of this). A more applicable way is to redefine it. However, if we redefine it without deleting it in advance, a warning will show up.

```
WARNING: Implicitly replacing the Component attribute c (type=<class
    'pyomo.core.base.constraint.SimpleConstraint'>) on block unknown with a
    new Component (type=<class
    'pyomo.core.base.constraint.AbstractSimpleConstraint'>). This is usually
    indicative of a modelling error. To avoid this warning, use
    block.del_component() and block.add_component().
```
However, for `Objective` in Pyomo, we can change its expression after it is defined.

bernalde replied:
I understand, alright. Let's keep going this way.

bernalde commented on pyomo/contrib/mindtpy/initialization.py:
```diff
@@ -153,7 +147,7 @@ def init_rNLP(solve_data, config):
         config.logger.info(
             'Relaxed NLP: OBJ: %s  LB: %s  UB: %s'
-            % (value(main_objective.expr), solve_data.LB, solve_data.UB))
+            % (value(MindtPy.objective_list[-1].expr), solve_data.LB, solve_data.UB))
```
For clarity, I would define a variable being `MindtPy.objective_list[-1]` with a correct name (`main_objective`)?

ZedongPeng replied:
OK

bernalde commented on pyomo/contrib/mindtpy/mip_solve.py:
```diff
@@ -298,17 +287,17 @@ def handle_master_max_timelimit(master_mip, solve_data, config):
+    if solve_data.objective_sense == minimize:
         solve_data.LB = max(
-            value(MindtPy.obj.expr), solve_data.LB)
+            value(MindtPy.MindtPy_oa_obj.expr), solve_data.LB)
```
I'm in favor of changing `MindtPy_oa_obj` to `mip_obj` since MindtPy already appears in the object name and we implement many methods besides OA.

ZedongPeng replied:
`mip_obj` or `master_obj`? We just replaced all `mip` with `master` several weeks ago.

bernalde replied:
I would go for `mip_obj`

bernalde commented on pyomo/contrib/mindtpy/util.py:
```diff
@@ -309,9 +301,7 @@ def generate_lag_objective_function(model, setpoint_model, config, discrete_only
+    temp_model.MindtPy_utils.objective_list[0].activate()
     temp_model.MindtPy_utils.deactivate()
     TransformationFactory('core.relax_integer_vars').apply_to(temp_model)
     # TODO: PyNumero does not support discrete variables
```
This is note a note than a TODO, right? You are handling it by the relax integrality transformation.

ZedongPeng replied:
Yes

bernalde reviewed (commented):
Hey Zedong, this was a big commit of clean-up, and it was really useful to help maintain the code. ALong those lines, I would suggest you to start merging master into this branch (they have been doing BIG changes to Pyomo recently and it is better if we address these early). Our goal is to have the tests passing.
This would also be a good time to rethink (if necessary) the way that we organize the info inside MindtPy. Seeing things like `MindtPy.MindtPy_linear_cuts...`, `fixed_nlp.MindtPy_utils.objective_list`, and `MindtPy.MindtPy_oa_obj.expr` make me think that we can clean this up a little bit. What are your thoughts?

ZedongPeng committed:
change TODO to note

ZedongPeng committed:
Merge branch 'master' into level-based-outer-approximation

bernalde commented:
The linux 3.9 slim test fails as 
```
  File "/home/runner/work/pyomo/pyomo/pyomo/contrib/mindtpy/plugins.py", line 12, in load
    import pyomo.contrib.mindtpy.MindtPy
  File "/home/runner/work/pyomo/pyomo/pyomo/contrib/mindtpy/MindtPy.py", line 38, in <module>
    from pyomo.contrib.mindtpy.initialization import MindtPy_initialize_master
  File "/home/runner/work/pyomo/pyomo/pyomo/contrib/mindtpy/initialization.py", line 16, in <module>
    from pyomo.contrib.mindtpy.nlp_solve import solve_subproblem
  File "/home/runner/work/pyomo/pyomo/pyomo/contrib/mindtpy/nlp_solve.py", line 17, in <module>
    from pyomo.contrib.mindtpy.util import add_feas_slacks
  File "/home/runner/work/pyomo/pyomo/pyomo/contrib/mindtpy/util.py", line 22, in <module>
    from pyomo.contrib.pynumero.interfaces.pyomo_nlp import PyomoNLP
  File "/home/runner/work/pyomo/pyomo/pyomo/contrib/pynumero/interfaces/pyomo_nlp.py", line 16, in <module>
    import numpy as np
```
I think this is because numpy is not available in that server and the numpy_available of PyNumero is bypassed when we import pyomo_nlp directly. We need to address this by using numpy_available inside of MindtPy.

bernalde commented:
We are facing again the ASCII character issue
`Original SyntaxError: Non-ASCII character '\xef' in file /Users/runner/work/pyomo/pyomo/pyomo/contrib/mindtpy/initialization.py on line 169, but no encoding declared; see http://python.org/dev/peps/pep-0263/ for details (initialization.py, line 169)`
I think we should find a more sustainable way of dealing with this than adding the comment at the beginning of the file. Maybe having the editor identify non-ASCII characters and replacing them could make the trick.
Check this webpage that gives a command for printing the problematic characters and see if we can remove them for good https://linuxconfig.org/syntaxerror-non-ascii-character-python-with-utf-8-encoding

ZedongPeng committed:
rename MindtPy_oa_obj to mip_obj

ZedongPeng committed:
add option calculate_dual

ZedongPeng committed:
add main_objective for clarity

ZedongPeng committed:
fix cycling check

ZedongPeng committed:
add doc for norm constraint in FP

ZedongPeng committed:
add doc for two implementations in generate_lag_objective_function

ZedongPeng committed:
rename obj_limit to obj_reg_estimate

ZedongPeng committed:
add warmstart for gams
